src/Storages/MergeTree/ZooKeeperWithFaultInjection.h

```diff
@@ -0,0 +1,527 @@
+        {
+            fault_cleanup();
+            throw;
+        }
+    }
+    void faultInjectionAfter(std::function<void()> fault_cleanup)
+    {
+        try
+        {
+            if (unlikely(fault_policy))
+                fault_policy->afterOperation();
+        }
+        catch (const zkutil::KeeperException &)
+        {
+            fault_cleanup();
+            throw;
+        }
+    }
+
+    void doForEachCreatedEphemeralNode(
+        const char * method, const Coordination::Requests & requests, const Coordination::Responses & responses, auto && action)
+    {
+        if (responses.empty())
+            return;
+
+        if (responses.size() != requests.size())
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "Number of responses doesn't match number of requests: method={} requests={} responses={}",
+                method,
+                requests.size(),
+                responses.size());
+
+        /// find create request with ephemeral flag
+        std::vector<std::pair<size_t, const Coordination::CreateRequest *>> create_requests;
+        for (size_t i = 0; i < requests.size(); ++i)
+        {
+            const auto * create_req = dynamic_cast<const Coordination::CreateRequest *>(requests[i].get());
+            if (create_req && create_req->is_ephemeral)
+                create_requests.emplace_back(i, create_req);
+        }
+
+        for (auto && [i, req] : create_requests)
+        {
+            const auto * create_resp = dynamic_cast<const Coordination::CreateResponse *>(responses.at(i).get());
+            if (!create_resp)
+                throw Exception(
+                    ErrorCodes::LOGICAL_ERROR, "Response should be CreateResponse: method={} index={} path={}", method, i, req->path);
+
+            action(create_resp->path_created);
+        }
+    }
+
+    void faultInjectionPostAction(const char * method, const Coordination::Requests & requests, Coordination::Responses & responses)
+    {
+        doForEachCreatedEphemeralNode(method, requests, responses, [&](const String & path_created) { keeper->remove(path_created); });
+    }
+
+    template <typename T>
+    struct FaultCleanupTypeImpl
+    {
+        using Type = std::function<void(T &)>;
+    };
+
+    template <>
+    struct FaultCleanupTypeImpl<void>
+    {
+        using Type = std::function<void()>;
+    };
+
+    template <typename T>
+    using FaultCleanupType = typename FaultCleanupTypeImpl<T>::Type;
+
+    template <
+        bool no_throw_access = false,
+        bool inject_failure_before_op = true,
+        int inject_failure_after_op = true,
+        typename Operation,
+        typename Result = std::invoke_result_t<Operation>>
+    Result access(
+        const char * func_name,
+        const std::string & path,
+        Operation operation,
+        FaultCleanupType<Result> fault_after_op_cleanup = {},
+        FaultCleanupType<void> fault_before_op_cleanup = {})
```

Comment:
```suggestion
        FaultCleanupType<Result> fault_after_op_cleanup = {},
        FaultCleanupType<void> fault_before_op_cleanup = {})
```

Why do you have 2 arguments?
Also, with this approach you don't need FaultCleanupType, first can be directly `std::function<void(Result &)>` and second `std::function<void()>`